runtime/parachains/src/scheduler.rs

```diff
@@ -386,70 +368,111 @@ impl<T: Config> Pallet<T> {
 	pub(crate) fn next_up_on_available(core: CoreIndex) -> Option<ScheduledCore> {
 		ClaimQueue::<T>::get().get(&core).and_then(|a| {
 			a.iter()
-				.position(|e| e.is_some())
-				.and_then(|pos| Self::assignment_to_scheduled_core(&a[pos]))
+				.find_map(|e| e.as_ref())
+				.map(|pe| Self::paras_entry_to_scheduled_core(pe))
 		})
 	}
 
-	fn assignment_to_scheduled_core(ca: &Option<CoreAssignment>) -> Option<ScheduledCore> {
-		match ca {
-			None => None,
-			Some(ca) => match ca.kind.clone() {
-				Assignment::Parachain(_para_id) =>
-					Some(ScheduledCore { para_id: ca.kind.para_id(), collator: None }),
-				Assignment::ParathreadA(entry) =>
-					Some(ScheduledCore { para_id: ca.kind.para_id(), collator: entry.claim.1 }),
-			},
+	fn paras_entry_to_scheduled_core(pe: &ParasEntry) -> ScheduledCore {
+		ScheduledCore {
+			para_id: pe.para_id(),
+			collator_restrictions: pe.collator_restrictions().clone(),
 		}
 	}
 
 	/// Return the next thing that will be scheduled on this core assuming it is currently
 	/// occupied and the candidate occupying it times out.
 	pub(crate) fn next_up_on_time_out(core: CoreIndex) -> Option<ScheduledCore> {
-		Self::next_up_on_available(core)
+		Self::next_up_on_available(core).or_else(|| {
+			// Or, if none, the claim currently occupying the core,
+			// as it would be put back on the queue after timing out if number of retries is not at the maximum.
+			let cores = AvailabilityCores::<T>::get();
+			cores.get(core.0 as usize).and_then(|c| match c {
+				CoreOccupied::Free => None,
+				CoreOccupied::Paras(pe) =>
+					if pe.retries < T::AssignmentProvider::get_max_retries(core) {
+						Some(Self::paras_entry_to_scheduled_core(pe))
+					} else {
+						None
+					},
+			})
+		})
+	}
+
+	/// Checks if `core_idx` occupied since `pending_since` will timeout after `after`.
+	fn timesout_after(
+		core_idx: CoreIndex,
+		pending_since: T::BlockNumber,
+		after: T::BlockNumber,
+	) -> bool {
+		let availability_period = T::AssignmentProvider::get_availability_period(core_idx);
+		after.saturating_sub(pending_since) >= availability_period
+	}
+
+	// on new session
+	/// `pending_cores` is a map from `CoreIndex` to `T::BlockNumber` which depicts when the `CoreIndex` was occupied.
+	fn push_occupied_cores_to_assignment_provider(
```

Comment:
While we only do this on new sessions or better right before them. I don't see with the current state, why this function should care. I think it should just do what the name suggests.